Google.Api.Gax.Grpc/Rest/PartialDecodingStreamReader.cs

```diff
@@ -0,0 +1,136 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net.Http;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Grpc.Core;
+using Newtonsoft.Json.Linq;
+
+namespace Google.Api.Gax.Grpc.Rest;
+
+/// <summary>
+/// An IAsyncStreamReader implementation that reads an array of messages
+/// from HTTP stream as they arrive in (partial) JSON chunks. 
+/// </summary>
+/// <typeparam name="TResponse">Type of proto messages in the stream</typeparam>
+public class PartialDecodingStreamReader<TResponse> : IAsyncStreamReader<TResponse>
+{
+    private readonly Func<string, TResponse> _parsing;
+
+    private readonly Queue<TResponse> _readyResults;
+    private readonly StringBuilder _currentBuffer;
+    
+    private readonly StreamReader _streamReader;
+    private bool _arrayClosed;
+
+    /// <summary>
+    /// Creates the StreamReader
+    /// </summary>
+    /// <param name="stream">A stream with partial JSON</param>
+    /// <param name="parsing">A function to transform a well-formed JSON object into the proto message.</param>
+    public PartialDecodingStreamReader(Stream stream, Func<string, TResponse> parsing)
+    {
+        _streamReader = new StreamReader(stream);
+        _parsing = parsing;
+        
+        _readyResults = new Queue<TResponse>();
+        _currentBuffer = new StringBuilder();
+        _arrayClosed = false;
+    }
+
+    /// <inheritdoc />
+    public async Task<bool> MoveNext(CancellationToken cancellationToken)
+    {
+        if (_readyResults.Count > 0)
+        {
+            Current = _readyResults.Dequeue();
+            return true;
+        }
+
+        while (_readyResults.Count == 0)
+        {
+            var buffer = new char[1000];
+            var taskRead = _streamReader.ReadAsync(buffer, 0, buffer.Length);
+            var cancellationTask = Task.Delay(-1, cancellationToken);
+            var resultTask = await Task.WhenAny(taskRead, cancellationTask).ConfigureAwait(false);
+
+            if (resultTask == cancellationTask)
+            {
+                // If the cancellationTask "wins" `Task.WhenAny` by being cancelled, the following await will throw TaskCancelledException.
+                await cancellationTask.ConfigureAwait(false);
+            }
+
+            var readLen = await taskRead.ConfigureAwait(false);
+            if (readLen == 0)
+            {
+                if (!_arrayClosed)
+                {
+                    var errorText = "Closing `]` bracket not received after iterating through the stream. " +
+                                    "This means that streaming ended without all objects transmitted. " +
+                                    "It is likely a result of server or network error.";
+                    throw new InvalidOperationException(errorText);
+                }
+
+                break;
```

Comment:
This "break" is effectively "return false" isn't it? I think it would be clearer to do that directly and remove the `if (!_readyResults.Any())` code later.

Reply:
👍